### PR TITLE
summer-ospp 2022: 飞桨PaddlePaddle Sparse Conv开发和优化: gather-gemm-scatter fuse

### PR DESCRIPTION
**Pull request by umiswing**
### PR types

Performance optimization

### PR changes

APIs, Others

### Describe

1. Include cutlass in Paddle, add ```cutlass.cmake``` to download cutlass v2.9.1 automatically. 
2. Fuse gather, gemm and scatter in ```Conv3dCOOGpuKernel``` with cutlass, this fusion supports ```fp16```, ```fp32``` and ```fp64```, but only works for channels cutlass supports. I remain the old code for other channels. 
3. Add ```gather_gemm_scatter.h``` and ```gather_gemm_scatter.cu```. These two files define the details of the fused kernel.

Following is a comparison of running 500 times forward with gather-gemm-grouped fusion and without this fusion on GeForce RTX 3060 Mobile GPU. This fusion can achieve 1.80 speedups at most. 
| time(seconds)\shape        | 0(subm)     | 1(subm)     | 2(conv)     | 3(subm)     | 4(conv)     | 5(subm)     | 6(conv)     | 7(subm)     | 8(subm)     | 9(conv)     |
| -------------------------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |
| fp16                       |             |             |             |             |             |             |             |             |             |             |
| fused                      | 0.948059366 | 0.948617946 | 0.918128433 | 0.654550546 | 1.247602738 | 0.643608378 | 3.765412761 | 0.467916381 | 0.469078923 | 0.166027756 |
| old                        | 0.963316033 | 0.978123623 | 1.100062921 | 0.85829027  | 1.841123472 | 1.073228994 | 6.164761058 | 0.813147195 | 0.815460634 | 0.211744664 |
| spconv 2.1.21              | 0.370279025 | 0.356675464 | 1.255860301 | 0.67564418  | 1.989892343 | 0.646101865 | 4.651363121 | 0.397059219 | 0.399020844 | 0.274251656 |
| old/fused                  | 1.016092523 | 1.031103857 | 1.198157993 | 1.311266602 | 1.475728945 | 1.667518682 | 1.637207246 | 1.737804505 | 1.738429492 | 1.275357019 |
| spconv/fused               | 0.390565231 | 0.375994852 | 1.367848174 | 1.032226135 | 1.594972728 | 1.00387423  | 1.235286386 | 0.848568751 | 0.850647566 | 1.651842214 |
| fp32                       |             |             |             |             |             |             |             |             |             |             |
| fused                      | 1.074077862 | 1.076842442 | 1.068928335 | 0.834602646 | 1.765838748 | 1.058910525 | 5.879794197 | 0.851036824 | 0.851216344 | 0.211274696 |
| old                        | 1.017575127 | 1.055912888 | 1.444613468 | 1.186929089 | 2.804123872 | 1.722058823 | 10.58918662 | 1.401214913 | 1.40421477  | 0.327167335 |
| spconv 2.1.21              | 0.367197311 | 0.379458345 | 1.375788394 | 0.765276265 | 2.479875851 | 1.204778016 | 7.822352023 | 0.972676926 | 0.973764882 | 0.298347718 |
| old/fused                  | 0.947394191 | 0.980563959 | 1.351459607 | 1.422148725 | 1.58798411  | 1.626255271 | 1.800945112 | 1.646479769 | 1.649656729 | 1.548540082 |
| spconv/fused               | 0.341872153 | 0.352380562 | 1.287072621 | 0.916934866 | 1.404361442 | 1.137752423 | 1.330378541 | 1.142931655 | 1.143968732 | 1.412131806 |
| fp64                       |             |             |             |             |             |             |             |             |             |             |
| fused                      | 1.420543399 | 1.208440223 | 2.657113268 | 2.489402152 | 9.968613399 | 9.779180488 | 65.87212427 | 9.246852072 | 9.22867435  | 1.489004803 |
| old                        | 1.544003572 | 1.587914612 | 3.242279789 | 3.420265195 | 11.49426685 | 11.72775161 | oom         | 11.08054339 | 11.07393647 | 1.80335558  |
| spconv 2.1.21(not support) |             |             |             |             |             |             |             |             |             |             |
| old/fused                  | 1.086910525 | 1.314019992 | 1.220226412 | 1.37393036  | 1.153045703 | 1.199257098 |             | 1.198304386 | 1.199948774 | 1.211114683 |

Shape details:
|      | (x, y, z)        | in_channels | out_channels | kernel_size | strides   | paddings  | nnz    |
| ---- | ---------------- | ----------- | ------------ | ----------- | --------- | --------- | ------ |
| 0    | (41, 1600, 1408) | 4           | 16           | (3, 3, 3)   | (1, 1, 1) | (0, 0, 0) | 136000 |
| 1    | (41, 1600, 1408) | 16          | 16           | (3, 3, 3)   | (1, 1, 1) | (0, 0, 0) | 136000 |
| 2    | (41, 1600, 1408) | 16          | 32           | (3, 3, 3)   | (2, 2, 2) | (1, 1, 1) | 136000 |
| 3    | (21, 800, 704)   | 32          | 32           | (3, 3, 3)   | (1, 1, 1) | (0, 0, 0) | 220939 |
| 4    | (21, 800, 704)   | 32          | 64           | (3, 3, 3)   | (2, 2, 2) | (1, 1, 1) | 220939 |
| 5    | (11, 400, 352)   | 64          | 64           | (3, 3, 3)   | (1, 1, 1) | (0, 0, 0) | 146376 |
| 6    | (11, 400, 352)   | 64          | 64           | (3, 3, 3)   | (1, 1, 1) | (0, 0, 0) | 146376 |
| 7    | (5, 200, 176)    | 64          | 64           | (3, 3, 3)   | (2, 2, 2) | (0, 1, 1) | 65421  |
| 8    | (5, 200, 176)    | 64          | 64           | (3, 3, 3)   | (1, 1, 1) | (0, 0, 0) | 65421  |
| 9    | (5, 200, 176)    | 64          | 64           | (3, 1, 1)   | (2, 1, 1) | (0, 0, 0) | 65421  |

### CODE REVIEW

**Comment by hwu36** [comment]
Haicheng from cutlass team here.  Happy to help.

File: paddle/phi/kernels/reduce_sum_kernel.cc [hunk 1]
```diff
@@ -87,3 +87,7 @@ PD_REGISTER_KERNEL(
 #if defined(PADDLE_WITH_XPU)
 PD_REGISTER_KERNEL(sum, XPU, ALL_LAYOUT, phi::SumKernel, float) {}
 #endif
+
+#if defined(PADDLE_WITH_XPU)
+PD_REGISTER_KERNEL(sum, XPU, ALL_LAYOUT, phi::SumKernel, float) {}
```
Comment (zkh2016) [inline paddle/phi/kernels/reduce_sum_kernel.cc]: 这里和88行重复了

File: paddle/phi/kernels/sparse/gpu/conv.cu.h [hunk 2]
```diff
@@ -763,5 +787,294 @@ int ProductRuleBook(const Context& dev_ctx,
...
+    std::cerr << "Failed to initialize CUTLASS Grouped GEMM kernel."
+              << std::endl;
+    return;
+  }
+
+  // Run the grouped GEMM object
+  status = gemm.run();
+
+  if (status != cutlass::Status::kSuccess) {
+    std::cerr << "Failed to run CUTLASS Grouped GEMM kernel." << std::endl;
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/conv.cu.h]: 这里可以用`VLOG`，需要`include <glog/logging.h>`

File: paddle/phi/kernels/sparse/gpu/conv.cu.h [hunk 3]
```diff
@@ -763,5 +787,294 @@ int ProductRuleBook(const Context& dev_ctx,
...
+  }
+
+  // Wait for completion
+  cudaError_t error = cudaDeviceSynchronize();
+
+  if (error != cudaSuccess) {
+    std::cerr << "Kernel execution error: " << cudaGetErrorString(error);
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/conv.cu.h]: 同上

File: paddle/phi/kernels/sparse/gpu/conv.cu.h [hunk 4]
```diff
@@ -14,9 +14,23 @@ limitations under the License. */
...
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
 #include <thrust/remove.h>
 #include <thrust/sort.h>
 #include <thrust/unique.h>
+
+#include <algorithm>
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/conv.cu.h]: 这几个头文件debug时候用的？看代码应该不需要可以去掉

File: paddle/phi/kernels/sparse/gpu/conv_kernel.cu [hunk 5]
```diff
@@ -164,31 +153,319 @@ void Conv3dCooGPUKernel(const GPUContext& dev_ctx,
   }
 
   const T* kernel_ptr = kernel.data<T>();
-  for (int i = 0; i < kernel_size; i++) {
-    if (h_counter_ptr[i] <= 0) {
-      continue;
+
+#ifdef PADDLE_WITH_CUTLASS
+  // currently, only support data type == fp32 and indices type == int32_t using
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/conv_kernel.cu]: 看已经支持多种dtype了，注释改下？

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -59,7 +59,7 @@ repos:
     -   id: cpplint-cpp-source
         name: cpplint
         description: Check C++ code style using cpplint.py.
-        entry: bash ./tools/codestyle/cpplint_pre_commit.hook
+        entry: bash ./tools/codestyle/cpplint_pre_commit.hook --exclude=paddle/phi/kernels/sparse/gpu/gather_gemm_grouped.h
```
Comment (zkh2016) [inline .pre-commit-config.yaml]: 为何这个文件不检查？

File: python/paddle/incubate/sparse/nn/functional/conv.py [hunk 7]
```diff
@@ -66,7 +67,7 @@ def _conv3d(x,
     if in_dynamic_mode():
         pre_bias = _C_ops.sparse_conv3d(x, weight, padding, dilation, stride,
                                         groups, subm,
-                                        key if key is not None else "")
+                                        key if key is not None else "", cutlass)
```
Comment (phlrain) [inline python/paddle/incubate/sparse/nn/functional/conv.py]: 不允许在参数表中使用和计算逻辑无关的参数

File: .pre-commit-config.yaml (hunk 6 above)
Comment (umiswing) [reply .pre-commit-config.yaml]: > 为何这个文件不检查？

cutlass提供的接口和cpplint的要求不一致，如果要对这个文件进行检查的话，需要修改cutlass此处的代码：https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/device_kernel.h#L57

File: paddle/phi/kernels/sparse/gpu/conv.cu.h [hunk 8]
```diff
@@ -763,5 +787,295 @@ int ProductRuleBook(const Context& dev_ctx,
...
+
+  // The code section below describes datatype for input, output matrices and
+  // computation between elements in input matrices.
+
+  // This code section describes whether you want to use tensor cores or regular
+  // SIMT cores on GPU SM
+  using MMAOp = cutlass::arch::OpClassTensorOp;
+
+  // This code section describes CUDA SM architecture number
+  using SmArch = cutlass::arch::Sm80;
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/conv.cu.h]: 在不同GPU上，这里可能会有问题？
可以试试根据__CUDA_ARCH__来选择不同的arch

File: paddle/phi/kernels/sparse/gpu/gather_gemm_grouped.h [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+  static int const kAlignmentB = MapArguments::kAlignmentB;
+  static int const kAlignmentC =
+      Epilogue::OutputTileIterator::kElementsPerAccess;
+
+  /// Warp count (concept: GemmShape)
+  using WarpCount = typename Mma::WarpCount;
+  static int const kThreadCount = 32 * WarpCount::kCount;
+
+  //
+  // Structures
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/gather_gemm_grouped.h]: 这个注释似乎没什么意义，可以去掉

File: paddle/phi/kernels/sparse/gpu/gather_gemm_grouped.h [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+  //
+
+  /// Argument structure
+  struct Arguments {
+    //
+    // Data members
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/gather_gemm_grouped.h]: 同上

File: paddle/phi/kernels/sparse/gpu/conv.cu.h [hunk 11]
```diff
@@ -763,5 +773,295 @@ int ProductRuleBook(const Context& dev_ctx,
...
+                                        // vectorized memory access. For half
+                                        // precision, it's 8 elements. This
+                                        // becomes the vector width of math
+                                        // instructions in epilogue too
+      ElementAccumulator,               // <- data type of accumulator
+      ElementComputeEpilogue>;  // <- data type for alpha in linear combination
+                                // function
+
+  // Number of pipelines you want to use
+  constexpr int NumStages = 3;
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/conv.cu.h]: 这里是否也需要根据gpu架构进行区分下？

**Comment by mnicely** [comment]
Very exciting to see CUTLASS in action. Wanted to let you 2.10 was released last month if you're interested in linking to it directly.

File: paddle/phi/kernels/sparse/gpu/conv_kernel.cu [hunk 12]
```diff
@@ -1,11 +1,8 @@
 /* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
-
```
Comment (zkh2016) [inline paddle/phi/kernels/sparse/gpu/conv_kernel.cu]: 这里应该是有空行的？

**Review by zkh2016 — DISMISSED** [review]
LGTM。
辛苦后面补充下反向融合的实现。

File: python/paddle/incubate/sparse/nn/functional/conv.py (hunk 7 above)
Comment (umiswing) [reply python/paddle/incubate/sparse/nn/functional/conv.py]: > 不允许在参数表中使用和计算逻辑无关的参数

好的，这个参数已经去掉了。

**Review by zkh2016 — APPROVED** [review]

**Pull request merged**